File: DailyChallenges/Scala/LeetCode_1162.scala
object Solution {
    def maxDistance(grid: Array[Array[Int]]): Int = {
        val n = grid.length
        val q = collection.mutable.Queue.empty[Int]
        for (i <- 0 until n; j <- 0 until n; if grid(i)(j) == 1)
            q.enqueue(i*100+j);
        
        val dir = Array(1, 0, -1, 0, 1)
        var steps = 0
        while (q.nonEmpty){
            steps += 1
            for (k <- 0 until q.length; num = q.dequeue; i = num/100; j = num%100;
                l <- 0 until 4; _i = i + dir(l) if _i>=0 && _i<n;
                _j = j + dir(l+1) if _j>=0 && _j<n; if grid(_i)(_j)==0){
                    grid(_i)(_j) = steps
                    q.enqueue(_i * 100 + _j)       
            }
        }
        if (steps <= 1) -1 else (steps-1)
    }
}

File: DailyChallenges/CPP/LeetCode_1162.cpp
class Solution {
public:
    int maxDistance(vector<vector<int>>& grid) {
        int n = grid.size();
        queue<int> q;
        for (int i=0; i<n; ++i)
            for (int j=0; j<n; ++j)
                if (grid[i][j] == 1)
                    q.push(i * 100 + j);

        if (q.empty()) return -1; // all water

        int dir[5] = {1, 0, -1, 0, 1};
        int steps = 0;
        while (!q.empty()){
            ++steps;
            int sz = q.size();
            for (int k = 0; k < sz; ++k){
                int i = q.front() / 100, j = q.front() % 100; q.pop();
                for (int k = 0; k<4; ++k){
                    int _i = i + dir[k], _j = j + dir[k+1];
                    if (_i>=0 && _i<n && _j>=0 && _j<n && grid[_i][_j] == 0){
                        grid[_i][_j] = steps;
                        q.push(_i * 100 + _j);
                    }
                }
            }
        }
        return (steps == 1) ? -1 : (steps-1);
    }
};
